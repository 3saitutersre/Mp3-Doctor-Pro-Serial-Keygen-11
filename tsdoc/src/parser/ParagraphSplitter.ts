// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { DocSection, type DocNode, DocNodeKind, DocParagraph, type DocPlainText } from '../nodes';

/**
 * The ParagraphSplitter is a secondary stage that runs after the NodeParser has constructed
 * the DocComment.  It splits DocParagraph nodes into multiple paragraphs by looking for
 * paragraph delimiters.  Following CommonMark conventions, paragraphs are delimited by
 * one or more blank lines.  (These lines end with SoftBreak nodes.)  The blank lines are
 * not discarded.  Instead, they are attached to the preceding paragraph.  If the DocParagraph
 * starts with blank lines, they are preserved to avoid creating a paragraph containing only
 * whitespace.
 */
export class ParagraphSplitter {
  private static readonly _whitespaceRegExp: RegExp = /^\s*$/;

  /**
   * Split all paragraphs belonging to the provided subtree.
   */
  public static splitParagraphs(node: DocNode): void {
    if (node instanceof DocSection) {
      ParagraphSplitter.splitParagraphsForSection(node);

      // (We don't recurse here, since sections cannot contain subsections)
    } else {
      for (const childNode of node.getChildNodes()) {
        ParagraphSplitter.splitParagraphs(childNode);
      }
    }
  }

  /**
   * Split all paragraphs belonging to the provided DocSection.
   */
  public static splitParagraphsForSection(docSection: DocSection): void {
    const inputNodes: ReadonlyArray<DocNode> = docSection.nodes;
    const outputNodes: DocNode[] = [];

    for (const oldNode of inputNodes) {
      if (oldNode.kind === DocNodeKind.Paragraph) {
        ParagraphSplitter._splitParagraph(oldNode as DocParagraph, outputNodes);
      } else {
        outputNodes.push(oldNode);
      }
    }

    // Replace the inputNodes with the outputNodes
    docSection.clearNodes();
    docSection.appendNodes(outputNodes);
  }

  private static _splitParagraph(oldParagraph: DocParagraph, outputNodes: DocNode[]): void {
    const inputParagraphNodes: ReadonlyArray<DocNode> = oldParagraph.nodes;

    let currentParagraph: DocParagraph = new DocParagraph({ configuration: oldParagraph.configuration });
    outputNodes.push(currentParagraph);

    const enum SplitterState {
      Start,
      AwaitingTrailer,
      ReadingTrailer
    }
    let state: SplitterState = SplitterState.Start;

    let currentIndex: number = 0;
    while (currentIndex < inputParagraphNodes.length) {
      // Scan forwards to the end of the line
      let isBlankLine: boolean = true;
      let lineEndIndex: number = currentIndex; // non-inclusive
      do {
        const node: DocNode = inputParagraphNodes[lineEndIndex++];
        if (node.kind === DocNodeKind.SoftBreak) {
          break;
        }
        if (isBlankLine) {
          if (!this._isWhitespace(node)) {
            isBlankLine = false;
          }
        }
      } while (lineEndIndex < inputParagraphNodes.length);

      // At this point, the line and SoftBreak will be in inputParagraphNodes.slice(currentIndex, lineEndIndex)

      switch (state) {
        case SplitterState.Start:
          // We're skipping any blank lines that start the first paragraph
          if (!isBlankLine) {
            state = SplitterState.AwaitingTrailer;
          }
          break;
        case SplitterState.AwaitingTrailer:
          // We already saw some content, so now we're looking for a blank line that starts the trailer
          // at the end of this paragraph
          if (isBlankLine) {
            state = SplitterState.ReadingTrailer;
          }
          break;
        case SplitterState.ReadingTrailer:
          // We already found the trailer, so now we're looking for a non-blank line that will
          // begin a new paragraph
          if (!isBlankLine) {
            // Start a new paragraph
            currentParagraph = new DocParagraph({ configuration: oldParagraph.configuration });
            outputNodes.push(currentParagraph);

            state = SplitterState.AwaitingTrailer;
          }
          break;
      }

      // Append the line onto the current paragraph
      for (let i: number = currentIndex; i < lineEndIndex; ++i) {
        currentParagraph.appendNode(inputParagraphNodes[i]);
      }

      currentIndex = lineEndIndex;
    }
  }

  private static _isWhitespace(node: DocNode): boolean {
    switch (node.kind) {
      case DocNodeKind.PlainText:
        const docPlainText: DocPlainText = node as DocPlainText;
        return ParagraphSplitter._whitespaceRegExp.test(docPlainText.text);
      default:
        return false;
    }
  }
}
