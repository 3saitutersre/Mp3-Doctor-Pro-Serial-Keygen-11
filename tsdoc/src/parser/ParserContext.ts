// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TextRange } from './TextRange';
import type { Token } from './Token';
import { DocComment } from '../nodes';
import type { TSDocConfiguration } from '../configuration/TSDocConfiguration';
import { ParserMessageLog } from './ParserMessageLog';

/**
 * An internal data structure that tracks all the state being built up by the various
 * parser stages.
 */
export class ParserContext {
  /**
   * The configuration that was provided for the TSDocParser.
   */
  public readonly configuration: TSDocConfiguration;

  /**
   * The `sourceRange` indicates the start and end of the original input that was parsed.
   */
  public readonly sourceRange: TextRange;

  /**
   * The text range starting from the opening `/**` and ending with
   * the closing `*\/` delimiter.
   */
  public commentRange: TextRange = TextRange.empty;

  /**
   * The text ranges corresponding to the lines of content inside the comment.
   */
  public lines: TextRange[] = [];

  /**
   * A complete list of all tokens that were extracted from the input lines.
   */
  public tokens: Token[] = [];

  /**
   * The parsed doc comment object.  This is the primary output of the parser.
   */
  public readonly docComment: DocComment;

  /**
   * A queryable log that reports warnings and error messages that occurred during parsing.
   */
  public readonly log: ParserMessageLog;

  public constructor(configuration: TSDocConfiguration, sourceRange: TextRange) {
    this.configuration = configuration;
    this.sourceRange = sourceRange;

    this.docComment = new DocComment({ configuration: this.configuration });

    this.log = new ParserMessageLog();
  }
}
