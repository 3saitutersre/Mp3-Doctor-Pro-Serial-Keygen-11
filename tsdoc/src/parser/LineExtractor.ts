import { TextRange } from './TextRange';
import { ParserContext } from './ParserContext';
import { TSDocMessageId } from './TSDocMessageId';

// Internal parser state
enum State {
  // Initial state, looking for "/*"
  BeginComment1,
  // Looking for "*" or "* " after "/*"
  BeginComment2,
  // Like State.CollectingLine except immediately after the "/**"
  CollectingFirstLine,
  // Collecting characters until we reach a newline
  CollectingLine,
  // After a newline, looking for the "*" that begins a new line, or the "*/" to end the comment
  AdvancingLine,
  // Exiting the parser loop
  Done
}

/**
 * The main API for parsing TSDoc comments.
 */
export class LineExtractor {
  private static readonly _whitespaceCharacterRegExp: RegExp = /^\s$/;

  /**
   * This step parses an entire code comment from slash-star-star until star-slash,
   * and extracts the content lines.  The lines are stored in IDocCommentParameters.lines
   * and the overall text range is assigned to IDocCommentParameters.range.
   */
  public static extract(parserContext: ParserContext): boolean {
    const range: TextRange = parserContext.sourceRange;
    const buffer: string = range.buffer;

    let commentRangeStart: number = 0;
    let commentRangeEnd: number = 0;

    // These must be set before entering CollectingFirstLine, CollectingLine, or AdvancingLine
    let collectingLineStart: number = 0;
    let collectingLineEnd: number = 0;

    let nextIndex: number = range.pos;
    let state: State = State.BeginComment1;

    const lines: TextRange[] = [];

    while (state !== State.Done) {
      if (nextIndex >= range.end) {
        // reached the end of the input
        switch (state) {
          case State.BeginComment1:
          case State.BeginComment2:
            parserContext.log.addMessageForTextRange(
              TSDocMessageId.CommentNotFound,
              'Expecting a "/**" comment', range);
            return false;
          default:
            parserContext.log.addMessageForTextRange(
              TSDocMessageId.CommentMissingClosingDelimiter,
              'Unexpected end of input', range);
            return false;
        }
      }

      const current: string = buffer[nextIndex];
      const currentIndex: number = nextIndex;
      ++nextIndex;
      const next: string = nextIndex < range.end ? buffer[nextIndex] : '';

      switch (state) {
        case State.BeginComment1:
          if (current === '/' && next === '*') {
            commentRangeStart = currentIndex;
            ++nextIndex; // skip the star
            state = State.BeginComment2;
          } else if (!LineExtractor._whitespaceCharacterRegExp.test(current)) {
            parserContext.log.addMessageForTextRange(
              TSDocMessageId.CommentOpeningDelimiterSyntax,
              'Expecting a leading "/**"',
              range.getNewRange(currentIndex, currentIndex + 1));
            return false;
          }
          break;
        case State.BeginComment2:
          if (current === '*') {
            if (next === ' ') {
              ++nextIndex; // Discard the space after the star
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.CollectingFirstLine;
          } else {
            parserContext.log.addMessageForTextRange(
              TSDocMessageId.CommentOpeningDelimiterSyntax,
              'Expecting a leading "/**"',
              range.getNewRange(currentIndex, currentIndex + 1));
            return false;
          }
          break;
        case State.CollectingFirstLine:
        case State.CollectingLine:
          if (current === '\n') {
            // Ignore an empty line if it is immediately after the "/**"
            if (state !== State.CollectingFirstLine || collectingLineEnd > collectingLineStart) {
              // Record the line that we collected
              lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = nextIndex;
            collectingLineEnd = nextIndex;
            state = State.AdvancingLine;
          } else if (current === '*' && next === '/') {
            if (collectingLineEnd > collectingLineStart) {
              lines.push(range.getNewRange(collectingLineStart, collectingLineEnd));
            }
            collectingLineStart = 0;
            collectingLineEnd = 0;
            ++nextIndex; // skip the slash
            commentRangeEnd = nextIndex;
            state = State.Done;
          } else if (!LineExtractor._whitespaceCharacterRegExp.test(current)) {
            collectingLineEnd = nextIndex;
          }
          break;
        case State.AdvancingLine:
          if (current === '*') {
            if (next === '/') {
              collectingLineStart = 0;
              collectingLineEnd = 0;

              ++nextIndex; // skip the slash
              commentRangeEnd = nextIndex;
              state = State.Done;
            } else {
              // Discard the "*" at the start of a line

              if (next === ' ') {
                ++nextIndex; // Discard the space after the star
              }

              collectingLineStart = nextIndex;
              collectingLineEnd = nextIndex;
              state = State.CollectingLine;
            }
          } else if (current === '\n') {
            // Blank line
            lines.push(range.getNewRange(currentIndex, currentIndex));
            collectingLineStart = nextIndex;
          } else if (!LineExtractor._whitespaceCharacterRegExp.test(current)) {
            // If the star is missing, then start the line here
            // Example: "/**\nL1*/"

            // (collectingLineStart was the start of this line)
            collectingLineEnd = nextIndex;
            state = State.CollectingLine;
          }
          break;
      }
    }

    /**
     * Only fill in these if we successfully scanned a comment
     */
    parserContext.commentRange = range.getNewRange(commentRangeStart, commentRangeEnd);
    parserContext.lines = lines;
    return true;
  }
}
