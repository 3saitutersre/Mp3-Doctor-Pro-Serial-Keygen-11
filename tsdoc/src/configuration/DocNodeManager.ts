import { DocNode } from '../nodes/DocNode';
import { StringChecks } from '../parser/StringChecks';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type DocNodeConstructor = new (...args: any[]) => DocNode;

export interface IDocNodeDefinition {
  docNodeKind: string;
  constructor: DocNodeConstructor;
}

interface IRegisteredDocNodeDefinition {
  docNodeKind: string;
  constructor: DocNodeConstructor;
  packageName: string;
  allowedChildKinds: Set<string>;
}

/**
 * Part of the {@link TSDocConfiguration} object.
 *
 * @remarks
 * If you define your own custom subclasses of `DocNode`, they must be registered with the `DocNodeManager`.
 * Use {@link DocNodeManager.registerAllowableChildren} to specify which {@link DocNodeContainer} subclasses
 * are allowed to contain your nodes.
 */
export class DocNodeManager {
  // Matches an ASCII TypeScript-style identifier.
  //
  // Example: "_myIdentifier99"
  private static readonly _nodeKindRegExp: RegExp = /^[_a-z][_a-z0-9]*$/i;

  private readonly _docNodeDefinitionsByKind: Map<string, IRegisteredDocNodeDefinition> = new Map<
    string,
    IRegisteredDocNodeDefinition
  >();
  private readonly _docNodeDefinitionsByConstructor: Map<
    DocNodeConstructor,
    IRegisteredDocNodeDefinition
  > = new Map<DocNodeConstructor, IRegisteredDocNodeDefinition>();

  /**
   * Registers a list of {@link IDocNodeDefinition} objects to be used with the associated
   * {@link TSDocConfiguration} object.
   */
  public registerDocNodes(packageName: string, definitions: ReadonlyArray<IDocNodeDefinition>): void {
    const packageNameError: string | undefined = StringChecks.explainIfInvalidPackageName(packageName);
    if (packageNameError) {
      throw new Error('Invalid NPM package name: ' + packageNameError);
    }

    for (const definition of definitions) {
      if (!DocNodeManager._nodeKindRegExp.test(definition.docNodeKind)) {
        throw new Error(
          `The DocNode kind ${JSON.stringify(definition.docNodeKind)} is not a valid identifier.` +
            ` It must start with an underscore or letter, and be comprised of letters, numbers, and underscores`
        );
      }

      let existingDefinition: IRegisteredDocNodeDefinition | undefined = this._docNodeDefinitionsByKind.get(
        definition.docNodeKind
      );

      if (existingDefinition !== undefined) {
        throw new Error(
          `The DocNode kind "${definition.docNodeKind}" was already registered` +
            ` by ${existingDefinition.packageName}`
        );
      }

      existingDefinition = this._docNodeDefinitionsByConstructor.get(definition.constructor);
      if (existingDefinition !== undefined) {
        throw new Error(
          `This DocNode constructor was already registered by ${existingDefinition.packageName}` +
            ` as ${existingDefinition.docNodeKind}`
        );
      }

      const newDefinition: IRegisteredDocNodeDefinition = {
        docNodeKind: definition.docNodeKind,
        constructor: definition.constructor,
        packageName,
        allowedChildKinds: new Set<string>()
      };

      this._docNodeDefinitionsByKind.set(definition.docNodeKind, newDefinition);
      this._docNodeDefinitionsByConstructor.set(definition.constructor, newDefinition);
    }
  }

  /**
   * Reports an error if the specified DocNode kind has not been registered.
   */
  public throwIfNotRegisteredKind(docNodeKind: string): void {
    if (!this._docNodeDefinitionsByKind.has(docNodeKind)) {
      throw new Error(`The DocNode kind "${docNodeKind}" was not registered with this TSDocConfiguration`);
    }
  }

  /**
   * For the given parent DocNode kind, registers the specified DocNode kinds as being allowable children of
   * the parent.
   *
   * @remarks
   * To prevent mistakes, `DocNodeContainer` will report an error if you try to add node that was not registered
   * as an allowable child of the container.
   */
  public registerAllowableChildren(parentKind: string, childKinds: ReadonlyArray<string>): void {
    const parentDefinition: IRegisteredDocNodeDefinition = this._getDefinition(parentKind);

    for (const childKind of childKinds) {
      this._getDefinition(childKind);
      parentDefinition.allowedChildKinds.add(childKind);
    }
  }

  /**
   * Returns true if the specified DocNode kind has been registered as an allowable child of the specified
   * parent DocNode kind.
   */
  public isAllowedChild(parentKind: string, childKind: string): boolean {
    const parentDefinition: IRegisteredDocNodeDefinition = this._getDefinition(parentKind);
    return parentDefinition.allowedChildKinds.has(childKind);
  }

  private _getDefinition(docNodeKind: string): IRegisteredDocNodeDefinition {
    const definition: IRegisteredDocNodeDefinition | undefined = this._docNodeDefinitionsByKind.get(
      docNodeKind
    );
    if (definition === undefined) {
      throw new Error(`The DocNode kind "${docNodeKind}" was not registered with this TSDocConfiguration`);
    }
    return definition;
  }
}
