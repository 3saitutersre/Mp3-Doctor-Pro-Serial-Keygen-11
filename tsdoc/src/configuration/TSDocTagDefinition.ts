import { StringChecks } from '../parser/StringChecks';
import { Standardization } from '../details/Standardization';

/**
 * Determines the type of syntax for a TSDocTagDefinition
 */
export enum TSDocTagSyntaxKind {
  /**
   * The tag is intended to be an inline tag.  For example: `{@link}`.
   */
  InlineTag,

  /**
   * The tag is intended to be a block tag that starts a new documentation
   * section.  For example: `@remarks`
   */
  BlockTag,

  /**
   * The tag is intended to be a modifier tag whose presences indicates
   * an aspect of the associated API item.  For example: `@internal`
   */
  ModifierTag
}

/**
 * Constructor parameters for {@link TSDocTagDefinition}
 */
export interface ITSDocTagDefinitionParameters {
  tagName: string;
  syntaxKind: TSDocTagSyntaxKind;
  allowMultiple?: boolean;
}

/**
 * @internal
 */
export interface ITSDocTagDefinitionInternalParameters extends ITSDocTagDefinitionParameters {
  standardization: Standardization;
}

/**
 * Defines a TSDoc tag that will be understood by the TSDocParser.
 */
export class TSDocTagDefinition {
  /**
   * The TSDoc tag name.  TSDoc tag names start with an at-sign ("@") followed
   * by ASCII letters using "camelCase" capitalization.
   */
  public readonly tagName: string;

  /**
   * The TSDoc tag name in all capitals, which is used for performing
   * case-insensitive comparisons or lookups.
   */
  public readonly tagNameWithUpperCase: string;

  /**
   * Specifies the expected syntax for this tag.
   */
  public readonly syntaxKind: TSDocTagSyntaxKind;

  /**
   * Indicates the level of support expected from documentation tools that implement
   * the standard.
   */
  public readonly standardization: Standardization;

  /**
   * If true, then this TSDoc tag may appear multiple times in a doc comment.
   * By default, a tag may only appear once.
   */
  public readonly allowMultiple: boolean;

  public constructor(parameters: ITSDocTagDefinitionParameters) {
    StringChecks.validateTSDocTagName(parameters.tagName);
    this.tagName = parameters.tagName;
    this.tagNameWithUpperCase = parameters.tagName.toUpperCase();
    this.syntaxKind = parameters.syntaxKind;
    this.standardization =
      (parameters as ITSDocTagDefinitionInternalParameters).standardization || Standardization.None;
    this.allowMultiple = !!parameters.allowMultiple;
  }
}
