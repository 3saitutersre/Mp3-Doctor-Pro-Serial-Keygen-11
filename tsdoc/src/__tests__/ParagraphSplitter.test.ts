import { TestHelpers } from '../parser/__tests__/TestHelpers';
import { ParagraphSplitter } from '../parser/ParagraphSplitter';
import { DocSection, DocPlainText, DocSoftBreak, DocParagraph, DocBlockTag } from '../index';
import { TSDocConfiguration } from '../configuration/TSDocConfiguration';

test('01 Basic paragraph splitting', () => {
  TestHelpers.parseAndMatchDocCommentSnapshot(
    [
      '/**',
      ' *    ',
      ' * This is the',
      ' * first paragraph.',
      ' *   \t   ',
      ' *  ',
      ' *   \t   ',
      ' * This is the second paragraph.',
      ' *',
      ' * This is the third paragraph.',
      ' *',
      ' *   ',
      ' */'
    ].join('\n')
  );
});

test('02 Basic paragraph splitting in blocks', () => {
  TestHelpers.parseAndMatchDocCommentSnapshot(
    ['/**', ' * P1', ' * @remarks P2', ' *', ' * P3 @deprecated P4', ' *', ' * P5', ' */'].join('\n')
  );
});

test('03 Degenerate comment framing', () => {
  TestHelpers.parseAndMatchDocCommentSnapshot(
    ['/** line 1', ' * line 2', '', ' *   @public line 3*/'].join('\n')
  );
});

test('04 Degenerate manually constructed nodes', () => {
  const configuration: TSDocConfiguration = new TSDocConfiguration();

  const docSection: DocSection = new DocSection({ configuration }, [
    new DocParagraph({ configuration }, [
      new DocPlainText({ configuration, text: '  para 1 ' }),
      new DocSoftBreak({ configuration }),
      new DocPlainText({ configuration, text: '   ' }),
      new DocSoftBreak({ configuration }),
      new DocPlainText({ configuration, text: ' \t  ' }),
      new DocPlainText({ configuration, text: '   ' }),
      new DocBlockTag({ configuration, tagName: '@public' }),
      new DocPlainText({ configuration, text: '  para 2 ' }),
      new DocSoftBreak({ configuration }),
      new DocSoftBreak({ configuration }),
      new DocPlainText({ configuration, text: '  para 3  ' })
    ]),
    // Currently we do not discard empty paragraphs
    new DocParagraph({ configuration })
  ]);

  ParagraphSplitter.splitParagraphsForSection(docSection);
  expect(TestHelpers.getDocNodeSnapshot(docSection)).toMatchSnapshot();
});
