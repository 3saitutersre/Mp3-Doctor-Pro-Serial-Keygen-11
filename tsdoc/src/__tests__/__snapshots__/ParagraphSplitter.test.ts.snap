// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`01 Basic paragraph splitting 1`] = `
Object {
  "s00_lines": Array [
    "",
    "This is the",
    "first paragraph.",
    "",
    "",
    "",
    "This is the second paragraph.",
    "",
    "This is the third paragraph.",
    "",
    "",
  ],
  "s01_gaps": Array [],
  "s02_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "This is the",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "first paragraph.",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "This is the second paragraph.",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "This is the third paragraph.",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
    ],
  },
  "s03_remarksBlock": undefined,
  "s04_privateRemarksBlock": undefined,
  "s05_deprecatedBlock": undefined,
  "s06_paramBlocks": Array [],
  "s07_typeParamBlocks": Array [],
  "s08_returnsBlock": undefined,
  "s09_customBlocks": Array [],
  "s10_inheritDocTag": undefined,
  "s11_modifierTags": Array [],
  "s12_logMessages": Array [],
}
`;

exports[`02 Basic paragraph splitting in blocks 1`] = `
Object {
  "s00_lines": Array [
    "P1",
    "@remarks P2",
    "",
    "P3 @deprecated P4",
    "",
    "P5",
  ],
  "s01_gaps": Array [],
  "s02_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "P1",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
    ],
  },
  "s03_remarksBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodes": Array [
          Object {
            "kind": "Excerpt: BlockTag",
            "nodeExcerpt": "@remarks",
          },
        ],
      },
      Object {
        "kind": "Section",
        "nodes": Array [
          Object {
            "kind": "Paragraph",
            "nodes": Array [
              Object {
                "kind": "PlainText",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: PlainText",
                    "nodeExcerpt": " P2",
                  },
                ],
              },
              Object {
                "kind": "SoftBreak",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: SoftBreak",
                    "nodeExcerpt": "[n]",
                  },
                ],
              },
              Object {
                "kind": "SoftBreak",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: SoftBreak",
                    "nodeExcerpt": "[n]",
                  },
                ],
              },
            ],
          },
          Object {
            "kind": "Paragraph",
            "nodes": Array [
              Object {
                "kind": "PlainText",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: PlainText",
                    "nodeExcerpt": "P3 ",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  "s04_privateRemarksBlock": undefined,
  "s05_deprecatedBlock": Object {
    "kind": "Block",
    "nodes": Array [
      Object {
        "kind": "BlockTag",
        "nodes": Array [
          Object {
            "kind": "Excerpt: BlockTag",
            "nodeExcerpt": "@deprecated",
          },
        ],
      },
      Object {
        "kind": "Section",
        "nodes": Array [
          Object {
            "kind": "Paragraph",
            "nodes": Array [
              Object {
                "kind": "PlainText",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: PlainText",
                    "nodeExcerpt": " P4",
                  },
                ],
              },
              Object {
                "kind": "SoftBreak",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: SoftBreak",
                    "nodeExcerpt": "[n]",
                  },
                ],
              },
              Object {
                "kind": "SoftBreak",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: SoftBreak",
                    "nodeExcerpt": "[n]",
                  },
                ],
              },
            ],
          },
          Object {
            "kind": "Paragraph",
            "nodes": Array [
              Object {
                "kind": "PlainText",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: PlainText",
                    "nodeExcerpt": "P5",
                  },
                ],
              },
              Object {
                "kind": "SoftBreak",
                "nodes": Array [
                  Object {
                    "kind": "Excerpt: SoftBreak",
                    "nodeExcerpt": "[n]",
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  "s06_paramBlocks": Array [],
  "s07_typeParamBlocks": Array [],
  "s08_returnsBlock": undefined,
  "s09_customBlocks": Array [],
  "s10_inheritDocTag": undefined,
  "s11_modifierTags": Array [],
  "s12_logMessages": Array [],
}
`;

exports[`03 Degenerate comment framing 1`] = `
Object {
  "s00_lines": Array [
    "line 1",
    "line 2",
    "",
    "  @public line 3",
  ],
  "s01_gaps": Array [],
  "s02_summarySection": Object {
    "kind": "Section",
    "nodes": Array [
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "line 1",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "line 2",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
      Object {
        "kind": "Paragraph",
        "nodes": Array [
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": "  ",
              },
            ],
          },
          Object {
            "kind": "PlainText",
            "nodes": Array [
              Object {
                "kind": "Excerpt: PlainText",
                "nodeExcerpt": " line 3",
              },
            ],
          },
          Object {
            "kind": "SoftBreak",
            "nodes": Array [
              Object {
                "kind": "Excerpt: SoftBreak",
                "nodeExcerpt": "[n]",
              },
            ],
          },
        ],
      },
    ],
  },
  "s03_remarksBlock": undefined,
  "s04_privateRemarksBlock": undefined,
  "s05_deprecatedBlock": undefined,
  "s06_paramBlocks": Array [],
  "s07_typeParamBlocks": Array [],
  "s08_returnsBlock": undefined,
  "s09_customBlocks": Array [],
  "s10_inheritDocTag": undefined,
  "s11_modifierTags": Array [
    Object {
      "kind": "BlockTag",
      "nodes": Array [
        Object {
          "kind": "Excerpt: BlockTag",
          "nodeExcerpt": "@public",
        },
      ],
    },
  ],
  "s12_logMessages": Array [],
}
`;

exports[`04 Degenerate manually constructed nodes 1`] = `
Object {
  "kind": "Section",
  "nodes": Array [
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 1 ",
        },
        Object {
          "kind": "SoftBreak",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "   ",
        },
        Object {
          "kind": "SoftBreak",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": " [t]  ",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "   ",
        },
        Object {
          "kind": "BlockTag",
        },
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 2 ",
        },
        Object {
          "kind": "SoftBreak",
        },
        Object {
          "kind": "SoftBreak",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
      "nodes": Array [
        Object {
          "kind": "PlainText",
          "nodePlainText": "  para 3  ",
        },
      ],
    },
    Object {
      "kind": "Paragraph",
    },
  ],
}
`;
